utils/estimated_histogram.hh

```diff
@@ -47,13 +47,376 @@
+     * higher bits are based on log2(value/min)
+     *
+     * lower bits are based on the high 2 MSB (ignoring the leading 1).
+
+     * for example: 330 (101001010)
+     * higher bits: log2(330/128) = 1
+     * low bit MSB: 330 = 01 (the lower two bits out of the upper 3)
+     * So the index: 101 = 5
+     *
+     *
+     * About the min/max and number of buckets.
+     * ========================================
+     *
+     * For MIN and MAX choose numbers that are a power of 2.
+     * The number of buckets will determine the resolution that should also be a power of 2
+     * So the total number of bucket should be log2(MAX/MIN) * Resolution + 1
+     *
+     * Limitation: You must set the MIN value to be higher then the resolution.
+     * For example, for a 2 bits resolution MIN should be 2^2 = 4 or higher.
+     *
+     */
+    using histogram_type = pseudo_floatingpoint_histogram<MIN, MAX, NUM_BUCKETS>;
+
+    static constexpr uint8_t RANGES = pow2_rank_constexpr(MAX/MIN);
+    static constexpr uint8_t RESOLUTION = (NUM_BUCKETS - 1)/RANGES;
+    static constexpr uint8_t RESOLUTION_BITS = pow2_rank_constexpr(RESOLUTION);
+    static constexpr uint8_t BASESHIFT = (MIN == 0) ? 0 : pow2_rank_constexpr(MIN);
+    static constexpr uint64_t LOWER_BITS_MASK = (1 << RESOLUTION_BITS) - 1;
+
+    std::vector<uint64_t> buckets;
+
+    pseudo_floatingpoint_histogram() {
+        assert(BASESHIFT >= RESOLUTION);
+        buckets.resize(NUM_BUCKETS, 0);
+    }
+
+    /*!
+     * \brief Returns the bucket limit given the bucket id.
+     * The first and last bucket will always return the MIN and MAX respectively.
+     *
+     */
+    uint64_t get_bucket_limit(uint16_t bucket_id) const {
+        if (bucket_id == 0) {
+            return MIN;
+        }
+        if (bucket_id == NUM_BUCKETS - 1) {
+            return MAX;
+        }
+        int16_t range_id = (bucket_id >> RESOLUTION_BITS);
+        return (1 << (range_id + BASESHIFT)) +  ((bucket_id & LOWER_BITS_MASK) << (range_id + BASESHIFT - RESOLUTION_BITS));
+    }
+
+    /*!
+     * \brief Find the bucket index for a given value
+     * The position of a value that is lower or equal to Min will always be 0.
+     * The position of a value is that is higher or equal to MAX will always be NUM_BUCKETS - 1.
+     */
+    uint16_t find_bucket_pos(uint64_t val) const {
+        if (val >= MAX) {
+            return NUM_BUCKETS - 1;
+        }
+        if (val <= MIN) {
+            return 0;
+        }
+        uint16_t range_id = pow2_rank(val);
+        val >>= range_id - RESOLUTION_BITS; // leave the top most N+1 bits where N is the resolution.
+        return ((range_id - BASESHIFT) << RESOLUTION_BITS) + (val & LOWER_BITS_MASK);
+    }
+
+    /*!
+     * \brief returns a cumulative histogram.
+     *
+     * The metrics cumulative histogram uses upper bounds.
+     * The histogram.count serves as an infinite upper bound bucket
+     */
+    seastar::metrics::histogram get_histogram() const {
```

Comment:
I see your point, I'll make it out of line.
A free function has it's pros, but in practice, it would make anyone that uses a histogram (typically for reporting it) life harder, because they would now need an extra step in returning a result when registering a histogram metrics.
This is how registering a histogram looks like:
```
sm::make_histogram("write_latency", sm::description("The general write latency histogram"),
                    {storage_proxy_stats::current_scheduling_group_label()},
                    [this]{return estimated_write.get_histogram();}),
```

If it's going to be an external function, it means that whoever is using histogram will need to include the file that holds that function and call it.

Reply:
It's not too bad, an additional include, and it makes the new code standalone.